### ISSUE TITLE
remove common indentation from strings and so forth

### ISSUE THREAD

**Issue by nikomatsakis** [dada-lang/dada#51]
At https://dada-lang.org/dyn_tutorial/hello.html, the tutorial claims:

> Strings in Dada can spread over multiple lines. Leading and trailing whitespace is stripped by default, and we also remove any common indentation from each line.

But this is clearly not true. =) We should fix this! I imagine this would take place in the parser? Or perhaps the validation engine, when it lowers format strings. Somewhere like that.

**Comment by XFFXFF** [comment]
I would like to know what is common indentation? 4, 8 or other. Can you give some hints on how to judge it? @nikomatsakis

**Comment by nikomatsakis** [comment]
@XFFXFF I imagined something like this:

* If the string has no `\n`, leave it as and return immediately
* Otherwise, break the string into lines and filter out empty lines
* Find the common prefix P of all those lines
* Find the leading whitespace of that common prefix P
* Strip that whitespace from every line in the original string, apart from empty lines, which remain as empty
* Strip leading/trailing whitespace as well

Some test cases:

---

```
print("
    Foo
        Bar
")
```

gives `"Foo\n    Bar"`.

---

```
print("    Foo")
```

gives "    Foo"

---

```
print("
\t    Foo
    \tBar
")
```

where \t is a tab character gives 

```
\t    Foo
    \tBar
```

in other words, there is no common prefix when tabs and spaces are mixed inconsistently. Seems fine, nobody should use tabs anyway. ;)

---

Oh, I just realized something else. I think that escaping of `\n` and other things should take place *after* this normalization has been done. So e.g.

```
print("    foo\n    bar\n").await
```

yields "    foo\n    bar\n"

In any case, I don't think we support escapes in our strings yet, but that's another bug. :)
